src/events/anti-raid/guildMemberAdd.ts

```diff
@@ -0,0 +1,75 @@
+import { on } from 'node:events';
+import { Client, Events, type GuildMember } from 'discord.js';
+import i18next from 'i18next';
+import type { Redis } from 'ioredis';
+import { inject, injectable } from 'tsyringe';
+import type { Event } from '../../Event.js';
+import { checkUsername } from '../../functions/anti-raid/usernameCheck.js';
+import { Case, CaseAction, createCase } from '../../functions/cases/createCase.js';
+import { upsertCaseLog } from '../../functions/logging/upsertCaseLog.js';
+import { checkLogChannel } from '../../functions/settings/checkLogChannel.js';
+import { getGuildSetting, SettingsKeys } from '../../functions/settings/getGuildSetting.js';
+import { logger } from '../../logger.js';
+import { kRedis } from '../../tokens.js';
+
+@injectable()
+export default class implements Event {
+	public name = 'User check';
+
+	public event = Events.GuildMemberAdd;
+
+	public constructor(public readonly client: Client<true>, @inject(kRedis) public readonly redis: Redis) {}
+
+	public async execute(): Promise<void> {
+		for await (const [guildMember] of on(this.client, this.event) as AsyncIterableIterator<[GuildMember]>) {
+			try {
+				const badNameHit = checkUsername(guildMember.user.username);
+
+				if (badNameHit) {
+					const logChannel = await checkLogChannel(
+						guildMember.guild,
+						(await getGuildSetting(guildMember.guild.id, SettingsKeys.ModLogChannelId)) as string,
+					);
+					if (!logChannel) {
+						continue;
+					}
+
+					const locale = (await getGuildSetting(guildMember.guild.id, SettingsKeys.Locale)) as string;
+
+					await this.redis.setex(`guild:${guildMember.guild.id}:user:${guildMember.id}:ban`, 15, '');
```

Comment:
There's a lot of duplicated logic in here from the ban command which should be abstracted rather than selectively copied.

Reply:
What can I do to improve this part?

Reply:
My thought here was that all of the logging and Redis logic is being duplicated and so should be abstracted into some more generic methods to handle the whole ban cycle, including logging.

Reply:
Addition to above:
Meaning abstract the logic in some common function that accepts all the necessary arguments (or grabs the containers: redis, etc from tsyringe into the function) and use the common function in both places the same code is used.